src/ostree/ot-builtin-commit.c

```diff
@@ -447,6 +551,31 @@ ostree_builtin_commit (int argc, char **argv, OstreeCommandInvocation *invocatio
       goto out;
     }
 
+  if (opt_bind_collection_refs)
+    {
+      int i, out_i;
+
+      if (opt_bind_refs)
+        {
+          g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                               "Only one of --bind-ref and --bind-collection-ref can be specified");
+          goto out;
+        }
+
+      /* For backwards compat, also set the same plain ref bindings */
+      opt_bind_refs = g_new0 (char *, g_strv_length (opt_bind_collection_refs) + 1);
+      for (i = 0, out_i = 0; opt_bind_collection_refs[i] != NULL; i++)
+        {
+          g_autofree char *ref_binding = NULL;
+          if (!split_collection_ref_pair (opt_bind_collection_refs[i],
+                                          NULL, &ref_binding, error))
+            goto out;
+
+          if (!g_strv_contains ((const char * const *)opt_bind_refs, ref_binding))
+            opt_bind_refs[out_i++] = g_steal_pointer (&ref_binding);
```

Comment:
Rather than calilng `g_strv_contains()` here (and getting O(N^2) behaviour, although there are probably never enough bindings for it to matter), perhaps it would be better to sort `opt_binding_refs` once it’s built, and then iterate and pairwise compare for duplicates? That gives the added advantage that the list of bindings is always sorted before being inserted into the commit metadata, which makes later lookups against it easier and stops the argument order to `ostree commit` from affecting the binding order (in a kind of ‘Reproducible Builds’) way.

Maybe best to do this with all the sets of bindings.

Reply:
That breaks the current "algorithm" of using the first specified ref command line argument as the one to pick for the backwards compat key though.

Reply:
Maybe there could be a separate block of code which picks the backwards compat key (and does it as you say)? It would be neat to get reproducible builds and to avoid O(N^2) behaviour.